# Welcome to Week 02 Handson - Data Preprocessing #01
In this hands-on session, we will learn some basic data pre-processings, which include:
1. duplicated data handling,
2. missing value handling,
3. data transformation (scalling and converting 'categorical data' to 'numerical data'),
4. outliers removal (for data preprocessing).

## Read dataset "raw-flight-data.csv"

In [160]:
import pandas as pd
import numpy as np


# read csv file into 'df' dataframe
df = pd.read_csv('./raw-flight-data.csv', sep = ",")

# print some data rows
df.head()

DayofMonth  DayOfWeek Carrier  OriginAirportID  DestAirportID  DepDelay  \
0          19          5      DL            11433          13303      -3.0   
1          19          5      DL            14869          12478       0.0   
2          19          5      DL            14057          14869      -4.0   
3          19          5      DL            15016          11433      28.0   
4          19          5      DL            11193          12892      -6.0   

   ArrDelay  
0       1.0  
1      -8.0  
2     -15.0  
3      24.0  
4     -11.0

## Milestone 01 (M01)
The given dataset is still a 'raw dataset' with duplicated data and missing values,<br>
1. In M01, please delete the duplicated data (keep only one) and return a dataframe with no duplicated data. **Hint:** use pandas API to handle the duplicated data,<br>
2. Print how many duplicated data (that are removed). **Hint:** calculate the difference of row numbers, before and after duplicate removal.

In [161]:
# write you own code for M01 here

duplicated_rows = df[df.duplicated()]
print("(2) length of removed duplicated data: {}".format(len(duplicated_rows)))
df_no_dup = df.drop_duplicates()
print("(1) Dataframe with no duplicated data: ")
df_no_dup

(2) length of removed duplicated data: 22435
(1) Dataframe with no duplicated data: 


DayofMonth  DayOfWeek Carrier  OriginAirportID  DestAirportID  \
0                19          5      DL            11433          13303   
1                19          5      DL            14869          12478   
2                19          5      DL            14057          14869   
3                19          5      DL            15016          11433   
4                19          5      DL            11193          12892   
...             ...        ...     ...              ...            ...   
2719413          19          6      DL            13204          14869   
2719414          19          6      DL            10397          13495   
2719415          19          6      DL            10140          10397   
2719416          19          6      DL            10397          13244   
2719417          19          6      DL            10397          13487   

         DepDelay  ArrDelay  
0            -3.0       1.0  
1             0.0      -8.0  
2            -4.0     -15.0  
3            28.0      24.0  
4            -6.0     -11.0  
...           ...       ...  
2719413       1.0      -3.0  
2719414      -2.0      -3.0  
2719415      18.0       9.0  
2719416      10.0       7.0  
2719417       0.0     -13.0  

[2696983 rows x 7 columns]

## M02
In M02, we will handle data rows having missing values. **Note:** in the given dataset, the missing values are only in the columns of ```DepDelay``` and ```ArrDelay```.
1. First and naive approach is by deleting the data rows having missing value. From data in M01, use pandas API to remove data rows with 'missing value', with specifications: (i) column subset to be checked = ```DepDelay``` and ```ArrDelay```, (ii) delete the data rows with **at least one missing value** in the given subset in (i),
2. Print the number missing rows from (1),
3. Another approach to handle missing values is by filling those missing values by their corresponding mean values, most frequent values, interpolated values, etc. In this M02, fill the missing values in the columns of ```DepDelay``` and ```ArrDelay``` by their corresponding mean values.
<font color="red">
4. Perform 'drop missing value', similar to (1), to the result of (3). Calculate the difference of row numbers before and after. You should get '0' for this. 
</font>
5. In your opinion, what are the differences between those two techniques (delete missing values vs filling with mean values)? Which one do you prefer? Please explain your argument.

In [162]:
# write you own code for M02 here

# df_no_dup.isnull().sum()

print("(1) & (2)  for first and naive approach")
df_isnull = df_no_dup.DepDelay.isnull() | df_no_dup.ArrDelay.isnull()

print("Number of rows contain NA value in either column DepDelay or ArrDelay : {}".format(len(df_no_dup.loc[df_isnull])))

df_miss1 = df_no_dup.dropna(subset=["DepDelay", "ArrDelay"])
df_miss1

(1) & (2)  for first and naive approach
Number of rows contain NA value in either column DepDelay or ArrDelay : 23798


DayofMonth  DayOfWeek Carrier  OriginAirportID  DestAirportID  \
0                19          5      DL            11433          13303   
1                19          5      DL            14869          12478   
2                19          5      DL            14057          14869   
3                19          5      DL            15016          11433   
4                19          5      DL            11193          12892   
...             ...        ...     ...              ...            ...   
2719413          19          6      DL            13204          14869   
2719414          19          6      DL            10397          13495   
2719415          19          6      DL            10140          10397   
2719416          19          6      DL            10397          13244   
2719417          19          6      DL            10397          13487   

         DepDelay  ArrDelay  
0            -3.0       1.0  
1             0.0      -8.0  
2            -4.0     -15.0  
3            28.0      24.0  
4            -6.0     -11.0  
...           ...       ...  
2719413       1.0      -3.0  
2719414      -2.0      -3.0  
2719415      18.0       9.0  
2719416      10.0       7.0  
2719417       0.0     -13.0  

[2673185 rows x 7 columns]

In [163]:
print("(3)  filling those missing values by their corresponding mean values")
df_miss2 = df_no_dup.fillna({"DepDelay": df_no_dup.DepDelay.mean(), "ArrDelay": df_no_dup.ArrDelay.mean()})
print("mean of DepDelay: {} and ArrDelay: {} ".format(df_no_dup.DepDelay.mean(), df_no_dup.ArrDelay.mean()))
print("list of indices which both DepDelay and ArrDelay are null: {}, and more..."
      .format(list(df_no_dup[df_no_dup.DepDelay.isnull() & df_no_dup.ArrDelay.isnull()].index)[0:5]))
df_miss2.loc[650]

(3)  filling those missing values by their corresponding mean values
mean of DepDelay: 10.618575625454712 and ArrDelay: 6.7272897311633875 
list of indices which both DepDelay and ArrDelay are null: [650, 651, 2268, 2269, 2597], and more...


DayofMonth              19
DayOfWeek                5
Carrier                 DL
OriginAirportID      13487
DestAirportID        13342
DepDelay           10.6186
ArrDelay           6.72729
Name: 650, dtype: object

Write your answer here for M02.5 (**Note**: in case you are not familiar, this is a 'markdown' environment) <br>

<font color='green'>
<i>Delete rows having missing values</i> sangat cocok diterapkan apabila dalam row tertentu terdapat lebih dari setengah column atribut memiliki nilai NaN. Model akan lebih bagus dengan menghilangkan rows yang memiliki missing values ini. Namun berpeluang untuk menghapus banyak rows sehingga informasi berkurang, terlebih lagi pada situasi dimana persentasi jumlah rows yang harus dihapus sangat tinggi dibanding dengan total jumlah semua data.
<br>

<i>Filling missing values with mean</i> berguna apabila atribut column bersifat kontinu, dan cocok untuk dataset yang kecil. Namun cara ini kurang efektif. Jika diterapkan pada dataset yang besar, terlebih lagi pada dataset yang tidak simetris, dapat terjadi resiko penyimpangan yang tinggi pada tahap training. 

Pada dataset di eksperimen ini, <i>Delete rows having missing values</i> lebih cocok diaplikasikan. Jumlah rows yang memiliki missing values ialah 23798, yaitu hanya 0,882% dari seluruh jumlah dataset. <i>Filling missing values with mean</i> kurang cocok pada dataset ini, dikarenakan sifat skewness data sangat tinggi dan banyak titik anomali berdasarkan boxplot pada eksperimen sebelumnya.

</font>

## M03
There is a 'Carrier' column in the given dataset, which is a "categorical variable". 
1. To build a learning model, what we need is numerical data so that our data can be processed by the learning model. Please convert the "categorical valriable" in the column of "Carrier" to "numerical variable". E.g., given categorical variable values of [A, B, A, C], we expect to get [0, 1, 0, 2] or [1, 2, 1, 3]. Thus, a row with the column value of "A", its value should be replaced with "0" (if you use this --> [0, 1, 0, 2]) or "1" (if you use this --> [1, 2, 1, 3]).

In [164]:
# write you own code for M03 here

# df.Carrier.unique()
to_replace = {'Carrier': {'DL': 0, 'AA': 1, 'EV': 2, 'B6': 3, 
                          'F9': 4, 'FL': 5, 'HA': 6, 'MQ': 7, 
                          '9E': 8, 'AS': 9, 'US': 10,'UA': 11, 
                          'VX': 12, 'WN': 13, 'YV': 14, 'OO': 15}}

final_df = df_miss1.copy()
final_df.replace(to_replace, inplace=True)
final_df

DayofMonth  DayOfWeek  Carrier  OriginAirportID  DestAirportID  \
0                19          5        0            11433          13303   
1                19          5        0            14869          12478   
2                19          5        0            14057          14869   
3                19          5        0            15016          11433   
4                19          5        0            11193          12892   
...             ...        ...      ...              ...            ...   
2719413          19          6        0            13204          14869   
2719414          19          6        0            10397          13495   
2719415          19          6        0            10140          10397   
2719416          19          6        0            10397          13244   
2719417          19          6        0            10397          13487   

         DepDelay  ArrDelay  
0            -3.0       1.0  
1             0.0      -8.0  
2            -4.0     -15.0  
3            28.0      24.0  
4            -6.0     -11.0  
...           ...       ...  
2719413       1.0      -3.0  
2719414      -2.0      -3.0  
2719415      18.0       9.0  
2719416      10.0       7.0  
2719417       0.0     -13.0  

[2673185 rows x 7 columns]

## M04
In some learning methods, we may perform data transformations so that we can expect a better performance (accuracy). One of the popular data transformation is **scalling**. Please do scalling the data column of "DepDelay" to [-1, 1]. <br>
**Hint:** In this hands-on session, don't use libary to do scalling, we expect you use scalling formula of:<br><br>
$$ \widehat{X} = \frac{(X-X_{min}) \times (BA - BB)}{X_{max}-X_{min}} + BB,$$ <br>
Where:<br>
$
\begin{align}
    \widehat{X} &= \text{scalled value}\\
    X &= \text{value being scalled}\\
    X_{max}, X_{min} &= \text{max of } X \text{, min of }X\\
    BA &= \text{batas atas, untuk kasus kita adalah 1}\\
    BB &= \text{batas bawah, untuk kasus kita adalah -1}\\
\end{align}
$

In [165]:
# write you own code for M04 here  

def fit_transform(tmp_df, BA=1, BB=-1, columns=[]):
    Xmax = tmp_df.max()
    Xmin = tmp_df.min()
    return tmp_df.apply(lambda x: ((x-Xmin[x.name]) * (BA-BB) / (Xmax[x.name]-Xmin[x.name]) + BB) 
                        if x.name in columns else x)

final_df = fit_transform(final_df, columns=["DepDelay"])
final_df

DayofMonth  DayOfWeek  Carrier  OriginAirportID  DestAirportID  \
0                19          5        0            11433          13303   
1                19          5        0            14869          12478   
2                19          5        0            14057          14869   
3                19          5        0            15016          11433   
4                19          5        0            11193          12892   
...             ...        ...      ...              ...            ...   
2719413          19          6        0            13204          14869   
2719414          19          6        0            10397          13495   
2719415          19          6        0            10140          10397   
2719416          19          6        0            10397          13244   
2719417          19          6        0            10397          13487   

         DepDelay  ArrDelay  
0       -0.937695       1.0  
1       -0.934579      -8.0  
2       -0.938733     -15.0  
3       -0.905504      24.0  
4       -0.940810     -11.0  
...           ...       ...  
2719413 -0.933541      -3.0  
2719414 -0.936656      -3.0  
2719415 -0.915888       9.0  
2719416 -0.924195       7.0  
2719417 -0.934579     -13.0  

[2673185 rows x 7 columns]

## M05
In raw data, there may be outliers and they should be removed before we use our data to our learning model. There are several approaches and one of them is by using Interquartile Range (IQR). The IQR can be used to identify outliers by defining limits on the sample values that are: (i) a factor $k$ of the IQR **below** the 25th percentile (Q1), **or** (ii) a factor $k$ of the IQR **above** the 75th percentile (Q3). The common value for the factor $k$ is the value 1.5 (thus, the outliers condition is below $(Q1 - 1.5 * IQR)$ or above $(Q3 + 1.5 * IQR)$). A factor $k$ of 3 or more can be used to identify values that are extreme outliers (far outs).

In M05, use the column ```ArrDelay``` to filter out the outliers data with $k=1.5$ and one more $k$ value you think it's a good value to filter out the outliers. Please also: (i) print the value of $Q1, Q2, IQR, (Q1 - k * IQR), (Q3 + k * IQR)$ and <font color='red'>(ii) count how many outliers you successfully remove by using $k=1.5$ and the other $k$ you choose</font>.

In [166]:
# write you own code for M05 here

# final_df.ArrDelay.describe()

k = 1.5
quantiles = list(final_df.ArrDelay.quantile([0.25, 0.5, 0.75]))
q1, q2, q3 = quantiles[0], quantiles[1], quantiles[2]
iqr = q3 - q1

def count_outliers(k=1.5):
    in_calc_min = (q1 - k * iqr)
    in_calc_max = (q3 + k * iqr)
    return len(final_df[(final_df.ArrDelay < in_calc_min) | (final_df.ArrDelay > in_calc_max)])

calc_min = (q1 - k * iqr)
calc_max = (q3 + k * iqr)
print("(i) Q1={}, Q2={}, Q3={}, IQR={}, (Q1 - k * IQR)={}, (Q3 + k * IQR)={}". 
      format(q1, q2, q3, iqr, calc_min, calc_max))

print("(ii) k=1.5, jumlah outliers={}".format(count_outliers()))

choose_k = 10
print("(ii) k={}, jumlah outliers={}".format(choose_k, count_outliers(choose_k)))


(i) Q1=-11.0, Q2=-3.0, Q3=10.0, IQR=21.0, (Q1 - k * IQR)=-42.5, (Q3 + k * IQR)=41.5
(ii) k=1.5, jumlah outliers=254931
(ii) k=10, jumlah outliers=12251


# Submission 
Submit this ipynb file to course portal similar to what you have done in the Week 01 Handson.